Add new 1.3 overwrite and csr_tracking_tag features for get_csr

Wenxing-hou opened the pull request:
Ref: #2291

1. Add new 1.3 overwrite and csr_tracking_tag features for get_csr;
2. Add unit_test for  get_csr new features;

The code has passed the unit_test.

Zhiqiang520 commented on include/hal/library/responder/csrlib.h:
```diff
@@ -48,6 +48,48 @@ extern bool libspdm_gen_csr(uint32_t base_hash_algo, uint32_t base_asym_algo, bo
+ * Gen CSR, which is uesd for SPDM 1.3
+ *
+ * @param[in]      base_hash_algo        Indicates the signing algorithm.
+ * @param[in]      base_asym_algo        Indicates the hash algorithm.
```
` * @param[in]      base_hash_algo        Indicates the signing algorithm.`
` * @param[in]      base_asym_algo        Indicates the hash algorithm.`

the above two comments were mismatched, should be as following:
` * @param[in]      base_hash_algo        Indicates the hash algorithm.`
` * @param[in]      base_asym_algo        Indicates the signing algorithm.`

Wenxing-hou replied:
Thanks. I have changed the code.

Zhiqiang520 commented on unit_test/test_spdm_responder/csr.c:
```diff
@@ -1333,6 +1336,519 @@ void libspdm_test_responder_csr_case12(void **state)
+    assert_int_equal(spdm_response->header.request_response_code,
+                     SPDM_ERROR);
+    assert_int_equal(spdm_response->header.param1,
+                     SPDM_ERROR_CODE_UNEXPECTED_REQUEST);
```
I think here `SPDM_ERROR_CODE_UNEXPECTED_REQUEST` should be `SPDM_ERROR_CODE_INVALID_REQUEST`.
Based on the  https://github.com/DMTF/libspdm/issues/2291#issuecomment-1822340186
![image](https://github.com/DMTF/libspdm/assets/90754803/264ff54d-3e38-4fce-a69d-05ec0214253a)

Wenxing-hou replied:
Thanks. I have changed the returned error code.

Zhiqiang520 commented:
I didn't find the test case for the following highlight test scenario, could you help double-check?
Based on the https://github.com/DMTF/libspdm/issues/2291#issuecomment-1822340186
![image](https://github.com/DMTF/libspdm/assets/90754803/e6c88c86-92e1-4ced-a4d6-74fd1a619450)

Zhiqiang520 commented on os_stub/spdm_device_secret_lib_sample/lib.c:
```diff
@@ -798,6 +789,209 @@ bool libspdm_gen_csr(uint32_t base_hash_algo, uint32_t base_asym_algo, bool *nee
+    uint8_t *cached_csr;
+    size_t csr_buffer_size;
+    uint8_t rsp_csr_tracking_tag;
+    uint8_t available_csr_tracking_tag;
```
`uint8_t *csr_tracking_tag` at line 871
`uint8_t rsp_csr_tracking_tag;` at line 879
`uint8_t available_csr_tracking_tag;` at line 880
The above three variable names are similar, what is the difference? It is hard to understand.

Wenxing-hou replied:
Thanks. I changed the `csr_tracking_tag` to `req_csr_tracking_tag`.

Wenxing-hou commented:
> I didn't find the test case for the following highlight test scenario, could you help double-check? Based on the [#2291 (comment)](https://github.com/DMTF/libspdm/issues/2291#issuecomment-1822340186) ![image](https://private-user-images.githubusercontent.com/90754803/292835933-e6c88c86-92e1-4ced-a4d6-74fd1a619450.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM2NDE4NjYsIm5iZiI6MTcwMzY0MTU2NiwicGF0aCI6Ii85MDc1NDgwMy8yOTI4MzU5MzMtZTZjODhjODYtOTJlMS00Y2VkLWE0ZDYtNzRmZDFhNjE5NDUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI3VDAxNDYwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA2OWM1MDU0Nzc4ZTcwOTUxYzM5ZGM0YjQwNGVkNzg4ODBjMmU1NzlmNDUxM2Y2Y2YyYWE0YjI0N2ViM2Y2YTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wdO4oqjiX_kGnwuJOr-lEcUF6xBBe2n0YMy-gCP6x1E)

Thanks. I found the unit_test in `libspdm_test_responder_csr_case14`,
```
 * Expected Behavior: the first get_csr with csr_tracking_tag 0: responder return need reset and available csr_tracking_tag;
 *                    Afer reset, then send get_csr with csr_tracking_tag 0 six times: responder return need reset and available csr_tracking_tag;
 *                    Then send get_csr with csr_tracking_tag 0: responder return busy error;
```

Wenxing-hou commented:
And I have added the table in code comment: https://github.com/DMTF/libspdm/blob/55c06f8434f28c84787635c8e4c3778d17be6455/include/hal/library/responder/csrlib.h#L55-L65

steven-bellock commented on include/library/spdm_requester_lib.h:
```diff
@@ -681,6 +681,33 @@ libspdm_return_t libspdm_get_csr(void *spdm_context,
+ * @param  key_pair_id                The value of this field shall be the unique key pair number identifying the desired
+ *                                    asymmetric key pair to associate with SlotID .
+ **/
+libspdm_return_t libspdm_get_csr_ex(void * spdm_context,
```
This function should return the `CSRTrackingTag` when the Responder sends a `ResetRequired` error message.

Wenxing-hou replied:
Yes. I have added the returned csr_tracking_tag parameter and added the unit_test.

steven-bellock commented on include/hal/library/responder/csrlib.h:
```diff
@@ -48,6 +48,60 @@ extern bool libspdm_gen_csr(uint32_t base_hash_algo, uint32_t base_asym_algo, bo
+ *                              On output, csr_pointer is address for stored CSR.
+ *                              The csr_pointer address will be changed.
+ *
+ * @param[in]       is_device_cert_model  If true, the cert chain is DeviceCert model.
```
`is_device_cert_model` -> `req_cert_model`
This is also now a `uint8_t` since the Requester can specify `DeviceCert`, `AliasCert`, or `GenericCert`.

Wenxing-hou replied:
Thanks. I think maybe we can update this parameter when we enable this feature.

steven-bellock replied:
At least for the API definition, the feature is now. Otherwise we'll need a new API to convert the `bool` to `uint8_t`.

Wenxing-hou replied:
Thanks. I have changed `is_device_cert_model` to `req_cert_model`.
But I just added simple logic in the l`ibspdm_gen_csr_ex`, which need enhance in the future.
https://github.com/DMTF/libspdm/blob/819b8513b25751fa14722a715b518cf6bcdbee24/os_stub/spdm_device_secret_lib_sample/lib.c#L995-L999

steven-bellock commented on library/spdm_responder_lib/libspdm_rsp_csr.c:
```diff
@@ -162,17 +164,61 @@ libspdm_return_t libspdm_get_response_csr(libspdm_context_t *spdm_context,
+            (spdm_request->header.param2 & SPDM_GET_CSR_REQUEST_ATTRIBUTES_CSR_TRACKING_TAG_MASK) >>
+            SPDM_GET_CSR_REQUEST_ATTRIBUTES_CSR_TRACKING_TAG_OFFSET;
+
+        if (csr_tracking_tag > SPDM_MAX_CSR_TRACKING_TAG) {
```
This check is not needed as `CSRTrackingTag` is 3 bits and its value cannot be larger than 7.

Wenxing-hou replied:
Thanks. I have deleted this invalid check.

steven-bellock commented on library/spdm_responder_lib/libspdm_rsp_csr.c:
```diff
@@ -162,17 +164,61 @@ libspdm_return_t libspdm_get_response_csr(libspdm_context_t *spdm_context,
-            spdm_context,
-            SPDM_ERROR_CODE_UNEXPECTED_REQUEST, 0,
-            response_size, response);
+        if ((csr_tracking_tag == 0) &&
```
The Integrator needs a better way to tell libspdm that it has run out of `CSRTrackingTag`s. Maybe have `csr_tracking_tag` return `0xFF`.

Wenxing-hou replied:
I have made the `csr_tracking_tag` return 0xFF when the device is busy.

Wenxing-hou commented on library/spdm_requester_lib/libspdm_req_get_csr.c:
```diff
@@ -87,10 +92,16 @@ static libspdm_return_t libspdm_try_get_csr(libspdm_context_t *spdm_context,
     spdm_request_size = message_size - transport_header_size -
                         spdm_context->local_context.capability.transport_tail_size;
 
+    if (libspdm_get_connection_version(spdm_context) == SPDM_MESSAGE_VERSION_12) {
```
Thanks. I have changed this check location.

steven-bellock commented on library/spdm_requester_lib/libspdm_req_get_csr.c:
```diff
@@ -89,8 +102,8 @@ static libspdm_return_t libspdm_try_get_csr(libspdm_context_t *spdm_context,
-    spdm_request->header.param1 = 0;
-    spdm_request->header.param2 = 0;
+    spdm_request->header.param1 = key_pair_id;
+    spdm_request->header.param2 = request_attribute;
```
I filed https://github.com/DMTF/SPDM-WG/issues/3319.

Wenxing-hou replied:
I have just added the check:
https://github.com/DMTF/libspdm/blob/819b8513b25751fa14722a715b518cf6bcdbee24/library/spdm_requester_lib/libspdm_req_get_csr.c#L73-L76

steven-bellock replied:
For now let's consider `(MULTI_KEY_CONN_RSP == true) && (CertModel == 0))` to be legal. If the SPDM WG determines that to be illegal then we can add the check to libspdm later.

steven-bellock commented on library/spdm_requester_lib/libspdm_req_get_csr.c:
```diff
@@ -154,6 +167,11 @@ static libspdm_return_t libspdm_try_get_csr(libspdm_context_t *spdm_context,
         goto receive_done;
     }
     if (spdm_response->header.request_response_code == SPDM_ERROR) {
+        if ((spdm_response->header.param1 == SPDM_ERROR_CODE_RESET_REQUIRED) &&
```
SPDM version must also be >= 1.3.

Wenxing-hou replied:
Thanks. I have added the version check.

Wenxing-hou commented on library/spdm_requester_lib/libspdm_req_get_csr.c:
```diff
@@ -89,8 +102,8 @@ static libspdm_return_t libspdm_try_get_csr(libspdm_context_t *spdm_context,
     spdm_request->header.request_response_code = SPDM_GET_CSR;
-    spdm_request->header.param1 = 0;
-    spdm_request->header.param2 = 0;
+    spdm_request->header.param1 = key_pair_id;
```
Thanks. I have added the check:
https://github.com/DMTF/libspdm/blob/819b8513b25751fa14722a715b518cf6bcdbee24/library/spdm_requester_lib/libspdm_req_get_csr.c#L68-L71

steven-bellock commented on library/spdm_requester_lib/libspdm_req_get_csr.c:
```diff
@@ -52,6 +59,24 @@ static libspdm_return_t libspdm_try_get_csr(libspdm_context_t *spdm_context,
+    }
+
+    if (libspdm_get_connection_version(spdm_context) >= SPDM_MESSAGE_VERSION_13) {
+        if ((spdm_context->connection_info.multi_key_conn_rsp) ^ (key_pair_id != 0)) {
```
Since this is not really bitwise XORing I'd suggest
```
(spdm_context->connection_info.multi_key_conn_rsp) == (key_pair_id == 0)
```

Wenxing-hou replied:
Thanks. I have changed the code to `(spdm_context->connection_info.multi_key_conn_rsp) == (key_pair_id == 0)`.

steven-bellock commented on include/hal/library/responder/csrlib.h:
```diff
@@ -48,6 +48,59 @@ extern bool libspdm_gen_csr(uint32_t base_hash_algo, uint32_t base_asym_algo, bo
+ * @retval  true   Success.
+ * @retval  false  Failed to gen CSR.
+ **/
+extern bool libspdm_gen_csr_ex(uint32_t base_hash_algo, uint32_t base_asym_algo, bool *need_reset,
```
Requiring this function is a breaking change and will need a new macro `LIBSPDM_ENABLE_CAPABILITY_CSR_CAP_EX`. Feel free to put it in `libspdm_lib_config.h` and specify that is only needed if the endpoints supports SPDM 1.3.

Wenxing-hou replied:
I have added the L`IBSPDM_ENABLE_CAPABILITY_CSR_CAP_EX`.  Both `libspdm_gen_csr_ex` and `libspdm_get_csr_ex` are under this MACRO.

steven-bellock replied:
Finally, add `LIBSPDM_ENABLE_CAPABILITY_CSR_CAP_EX` to https://github.com/DMTF/libspdm/blob/53f1cf915ed7cfd40b6712c722e8ef140cd7606a/.github/workflows/build.yml#L50-L51.

steven-bellock commented on include/hal/library/responder/csrlib.h:
```diff
@@ -48,6 +48,59 @@ extern bool libspdm_gen_csr(uint32_t base_hash_algo, uint32_t base_asym_algo, bo
+ *
+ * @param[in]      base_hash_algo        Indicates the hash algorithm.
+ * @param[in]      base_asym_algo        Indicates the signing algorithm.
+ * @param[in]      need_reset            If true, then device needs to be reset to complete the CSR.
```
This is [in/out]. 

On input it gives the value of `CERT_INSTALL_RESET_CAP`.
On output it specifies whether the device needs to be reset to complete the CSR or not.

Wenxing-hou replied:
Thanks. I have clear the comment and updated all the [in, out] parameters.

Wenxing-hou commented on library/spdm_responder_lib/libspdm_rsp_csr.c:
```diff
@@ -162,17 +165,58 @@ libspdm_return_t libspdm_get_response_csr(libspdm_context_t *spdm_context,
-                             requester_info, requester_info_length,
-                             opaque_data, opaque_data_length,
-                             &csr_len, csr_p, is_device_cert_model);
+
```
I have added the check:
https://github.com/DMTF/libspdm/blob/5e17eb3c2aa467af06b55f440fa4cf3bde4ebc5a/library/spdm_responder_lib/libspdm_rsp_csr.c#L188-L194

And I moved the `Overwrite` and `CSRTrackingTag` check form `libspdm_gen_csr_ex` to here, which make the code clearer.

steven-bellock commented:
Will approve once `LIBSPDM_ENABLE_CAPABILITY_CSR_CAP_EX` is added to CI/CD.

Wenxing-hou commented:
> Will approve once `LIBSPDM_ENABLE_CAPABILITY_CSR_CAP_EX` is added to CI/CD.

Thanks. I have added the `LIBSPDM_ENABLE_CAPABILITY_CSR_CAP_EX`  to CI/CD. But I added the MACRO in the follow:
https://github.com/DMTF/libspdm/blob/53f1cf915ed7cfd40b6712c722e8ef140cd7606a/.github/workflows/build.yml#L52-L53

Because the `LIBSPDM_ENABLE_CAPABILITY_CSR_CAP_EX` is under the `LIBSPDM_ENABLE_CAPABILITY_CSR_CAP`, we need spilt them to cover all the case.
